### PR TITLE
Cache: dynamic cache with cross attention and UMT5 `Cache` support

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

#28065 was becoming messy due to all Bart "copied from" dependencies, so this PR is a tiny version of it.

This PR:
1. Introduces `DynamicCacheWithCrossAttention`, which expands `DynamicCache` [cache object equivalent to the previous `past_key_values` input/output] with the ability to hold a cross-attention cache. This design was intentional: most LLMs (and now even multimodel models) tend to be decoder-only, so this separation will keep the cache class for decoder-only models simpler. It also enables us to be more strict -- in #28065 I've caught an unintended cache deletion in Whisper thanks to the increased specificity!
2. Adds `Cache` support to `modeling_umt5.py`, which is a form to test whether `DynamicCacheWithCrossAttention` is equivalent to the previous cache. These changes are the equivalent of the modeling changes in #26681, but for encoder-decoder models.

______________________________________

Local tests run:
1. `RUN_SLOW=1 py.test tests/models/umt5/test_modeling_umt5.py -vv` [Note: adds a test to ensure we keep the same results as in `main`]

### CODE REVIEW

File: src/transformers/models/umt5/modeling_umt5.py [hunk 1]
```diff
@@ -240,41 +248,71 @@ def compute_bias(self, query_length, key_length, device=None):
         values = values.permute([2, 0, 1]).unsqueeze(0)  # shape (1, num_heads, query_length, key_length)
         return values
 
+    def _prepare_key_values(
```
Comment (gante) [inline src/transformers/models/umt5/modeling_umt5.py:251]: This abstraction does not look particularly useful here. However, for models with multiple attention implementations, this abstraction is useful: all attention implementations can share it! 

(e.g. in Bart the benefits are clear)

File: src/transformers/models/umt5/modeling_umt5.py [hunk 2]
```diff
@@ -481,6 +501,7 @@ class UMT5PreTrainedModel(PreTrainedModel):
     supports_gradient_checkpointing = True
     _no_split_modules = ["UMT5Block"]
     _keep_in_fp32_modules = ["wo"]
+    _supports_cache_class = True
```
Comment (gante) [inline src/transformers/models/umt5/modeling_umt5.py:504]: This enables the `test_new_cache_format` test -> converting back and forth between the new cache and the legacy cache with cross attention is tested

File: tests/models/umt5/test_modeling_umt5.py [hunk 3]
```diff
@@ -560,6 +560,27 @@ def test_training_gradient_checkpointing_use_reentrant_false(self):
 @require_sentencepiece
 @require_tokenizers
 class Umt5IntegrationTest(unittest.TestCase):
+    def test_generation(self):
```
Comment (gante) [inline tests/models/umt5/test_modeling_umt5.py:563]: Ensures there is no regression from `main` 

I've double-checked that we get the same values in `main`. I've also checked the results with and without cache, in both `main` and this PR.

File: src/transformers/models/umt5/modeling_umt5.py [hunk 4]
```diff
@@ -1310,21 +1346,34 @@ def prepare_inputs_for_generation(
...
+            # 2 - Otherwise (past_length >= decoder_input_ids.shape[1]), let's assume decoder_input_ids only has
+            # unprocessed tokens.
+
+            # If we are about to go beyond the maximum cache length, we need to crop the input attention mask.
+            if (
+                max_cache_length is not None
+                and decoder_attention_mask is not None
+                and cache_length + decoder_input_ids.shape[1] > max_cache_length
+            ):
+                decoder_attention_mask = decoder_attention_mask[:, -max_cache_length:]
```
Comment (gante) [inline src/transformers/models/umt5/modeling_umt5.py:1373]: logic copied from llama + sink cache -> this makes the model ready for caches like sink cache

**Review by ArthurZucker — COMMENTED** [review]
#27931  will shamble things up 👿

**Pull request closed without merging**